File: apps/backend/src/routes/pairing.ts
/**
 * Pairing codes: the normal way to add a phone. The desktop (already
 * logged in) mints a short-lived, single-use code; the phone redeems it
 * for a device token without ever typing an email or password.
 */
import type { FastifyInstance } from "fastify";

import { randomPairingCode, randomTokenHex, sha256Hex } from "../auth.js";
import { Errors } from "../errors.js";
import { requireAuth } from "../plugins/auth.js";
import type { AppState } from "../state.js";

interface RedeemPairingBody {
  code?: string;
  name?: string;
  platform?: string;
}

function validateDeviceName(name: string): string {
  const trimmed = name.trim();
  if (trimmed.length === 0 || [...trimmed].length > 80) {
    throw Errors.validation("Device name must be 1-80 characters");
  }
  return trimmed;
}

export function registerPairingRoutes(app: FastifyInstance, state: AppState): void {
  app.post("/v1/pairing/create", async (request, reply) => {
    const userId = await requireAuth(state, request, reply);
    if (!userId) return;
    const now = new Date();
    const expiresAt = new Date(now.getTime() + state.config.pairingCodeTtlSecs * 1000);
    // Astronomically unlikely to collide, but retry a couple of times
    // rather than surface an internal error to the user.
    for (let attempt = 0; attempt < 5; attempt++) {
      const code = randomPairingCode();
      try {
        await state.db.createPairingCode(code, userId, now, expiresAt);
        reply.send({ code, expiresAt: expiresAt.toISOString() });
        return;
      } catch (e) {
        if (e instanceof Error && e.message === "CODE_COLLISION") continue;
        throw e;
      }
    }
    throw Errors.internal("could not generate a unique pairing code");
  });

  app.post<{ Body: RedeemPairingBody }>("/v1/pairing/redeem", async (request, reply) => {
    const code = (request.body.code ?? "").trim().toUpperCase();
    const name = validateDeviceName(request.body.name ?? "");
    const userId = await state.db.redeemPairingCode(code, new Date());
    if (!userId) throw Errors.invalidPairingCode();
    const token = randomTokenHex();
    const device = await state.db.createDevice(crypto.randomUUID(), userId, name, "mobile", (request.body.platform ?? "").trim(), sha256Hex(token), new Date());
    request.log.info({ userId, deviceId: device.id }, "device paired via code");
    reply.send({ deviceId: device.id, deviceToken: token, userId });
  });
}

File: crates/job-hunter-relay/src/routes/pairing.rs
//! Pairing codes: the normal way to add a phone. The desktop (already
//! logged in) mints a short-lived, single-use code; the phone redeems it
//! for a device token without ever typing an email or password.

use axum::extract::State;
use axum::Json;
use chrono::Utc;
use serde::{Deserialize, Serialize};

use crate::auth::{random_pairing_code, random_token_hex, sha256_hex};
use crate::error::{RelayError, RelayResult};
use crate::extractor::AuthUser;
use crate::models::DeviceKind;
use crate::state::AppState;
use crate::util::new_id;

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CreatePairingResponse {
    pub code: String,
    pub expires_at: chrono::DateTime<Utc>,
}

pub async fn create_pairing_code(
    AuthUser(user_id): AuthUser,
    State(state): State<AppState>,
) -> RelayResult<Json<CreatePairingResponse>> {
    let now = Utc::now();
    let expires_at = now + chrono::Duration::seconds(state.config.pairing_code_ttl_secs);
    // Astronomically unlikely to collide, but retry a couple of times
    // rather than surface an internal error to the user.
    for _ in 0..5 {
        let code = random_pairing_code();
        match state
            .db
            .create_pairing_code(&code, &user_id, now, expires_at)
        {
            Ok(()) => return Ok(Json(CreatePairingResponse { code, expires_at })),
            Err(RelayError::Internal(_)) => continue,
            Err(e) => return Err(e),
        }
    }
    Err(RelayError::Internal(
        "could not generate a unique pairing code".into(),
    ))
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RedeemPairingRequest {
    pub code: String,
    pub name: String,
    #[serde(default)]
    pub platform: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RedeemPairingResponse {
    pub device_id: String,
    pub device_token: String,
    pub user_id: String,
}

pub async fn redeem_pairing_code(
    State(state): State<AppState>,
    Json(req): Json<RedeemPairingRequest>,
) -> RelayResult<Json<RedeemPairingResponse>> {
    let code = req.code.trim().to_uppercase();
    let name = req.name.trim();
    if name.is_empty() || name.chars().count() > 80 {
        return Err(RelayError::Validation(
            "Device name must be 1-80 characters".into(),
        ));
    }
    let user_id = state
        .db
        .redeem_pairing_code(&code, Utc::now())?
        .ok_or(RelayError::InvalidPairingCode)?;
    let token = random_token_hex();
    let device = state.db.create_device(
        &new_id(),
        &user_id,
        name,
        DeviceKind::Mobile,
        req.platform.trim(),
        &sha256_hex(&token),
        Utc::now(),
    )?;
    tracing::info!(user_id = %user_id, device_id = %device.id, "device paired via code");
    Ok(Json(RedeemPairingResponse {
        device_id: device.id,
        device_token: token,
        user_id,
    }))
}
